update duplicate post microcopy

jazzsequence opened the pull request:
Updates the various bits of microcopy across the Duplicate Post plugin
Some of this involved making changes to strings in the block editor, which necessitated adding a build process
Moved all the duplicate post functions into a new namespace to keep the plugin clean

roborourke reviewed (commented):
I think my main concern is taking on responsibility for the duplicated parts of the JS. I was wondering if all we really need is to run `wp.i18n.setLocaleData()` with an object of all the strings present in the JS replacing them with our own. We might not even need the build process stuff if that works eg:

```php
wp_add_inline_script( 
  'duplicate_post_strings', 
  sprintf(
    'wp.i18n.setLocaleData( %s, 'duplicate-post' ); window.duplicatePostStrings = Object.assign( duplicatePostStrings || {}, %s ); duplicatePostStrings.checkLink = ',
    wp_json_encode( [ 'Republish' => [ 'Amend' ], ... ] ),
    wp_json_encode( [ 'checkLink' => __( '....', 'altis-workflow' ) ] )
  ),
  'after'
);
```

PHP component looks good but made one or two suggestions that might make maintenance easier.

jazzsequence commented on inc/clone-amend/namespace.php:
```diff
@@ -0,0 +1,470 @@
+
+	$original_post_edit_link = get_edit_post_link( $original_post->ID );
+
+	if ( isset( $post_states[0] ) && $post_states[0] === 'Customization Draft' ) {
```
tbh I'm not sure where "Customization Draft" even came from. I will do a bit of sleuthing to see if I can track it down.

jazzsequence replied:
Found it. Apparently a core thing. `¯\_(ツ)_/¯`

jazzsequence commented:
I've broken the suggested inline script down to the basest element and hooked it into both `enqueue_block_editor_assets` and `admin_enqueue_scripts` and neither actually replaces the strings. I've done this with both the original strings (e.g. 'Publish', et al) as well as the Duplicate Post strings (e.g. 'Republish', et al).

```php
	wp_add_inline_script(
		'duplicate_post_strings',
		sprintf(
		  'wp.i18n.setLocaleData( %s, \'duplicate-post\' );',
		  wp_json_encode( [ 'Republish' => [ 'Amend' ], ] ),
		),
		'after'
	  );
```

Unless I'm doing something very wrong, I'm not sure there's a way around managing the JavaScript in-house. 
That said, while the plugin code might change, our code may not need to as long as the WordPress core internals don't change as it doesn't rely on anything from the plugin, specifically.

goldenapples commented:
The inline script doesn't look like the correct function signature for [`setLocaleData()`](https://github.com/WordPress/gutenberg/blob/1ca5f6c1fbd813d33f69f249ac732e9f1c0ba3b2/i18n/index.js#L8-L17), which expects a Jed configuration object. Why wouldn't this be the same format that you're using in https://github.com/humanmade/altis-workflow/pull/73/files#diff-7ba4ff34a43209b539a078a4b05a572206363ff4d0987edfa41573140c96dde9R130-R135? 

And I'm not sure looking at that if even that will work, because as I understand it `setLocaleData` overrides the entire localization object. I would think that you'd have to check the existing localization and merge the new strings into it before setting the locale data.

roborourke commented:
@goldenapples pretty sure it merges the objects as it adds strings it encounters just in time. `setLocaleData()` expects just the `locale_data` portion of the Jed config object, so strings as keys and an array of strings for the translations.

Made a follow up in #82 that works.

roborourke commented:
> The inline script doesn't look like the correct function signature for [`setLocaleData()`](https://github.com/WordPress/gutenberg/blob/1ca5f6c1fbd813d33f69f249ac732e9f1c0ba3b2/i18n/index.js#L8-L17), which expects a Jed configuration object. Why wouldn't this be the same format that you're using in https://github.com/humanmade/altis-workflow/pull/73/files#diff-7ba4ff34a43209b539a078a4b05a572206363ff4d0987edfa41573140c96dde9R130-R135?
> 
> And I'm not sure looking at that if even that will work, because as I understand it `setLocaleData` overrides the entire localization object. I would think that you'd have to check the existing localization and merge the new strings into it before setting the locale data.

Interesting, looks like the duplicate-post plugin that the code was copied from is using `setLocaleData()` incorrectly, the array of strings is used to provide plural versions of a string so the second item in the array `duplicate-post` isn't actually doing anything. The second param to `setLocaleData()` is the text domain you're setting strings for. In this instance is overriding the default text domain so it works.

jazzsequence commented:
Merged in #82 -- that solution was working for me when tested...guess the real thing that wasn't working was the script I was using to hook to when I was trying, which makes sense.

Added a new notice string that appears when you are editing the _original_ post that has been amended, which I realized hadn't been covered (there's a few other notices, like the bulk edit notice, but I think this covers most things) and I cleaned up some of the logic around those notice things because I realized there was an extra check that was making an unnecessary database query for `get_post_meta`.

roborourke commented:
🥳
